tremor-pipeline/src/query.rs

```diff
@@ -174,51 +183,38 @@ impl Query {
             .and_then(Value::as_str)
             .unwrap_or("<generated>");
 
-        // TODO compute public streams - do not hardcode
-        let id = pipe_graph.add_node(NodeConfig {
-            id: IN,
-            kind: NodeKind::Input,
-            op_type: "passthrough".to_string(),
-            ..NodeConfig::default()
-        });
-        nodes.insert(IN, id);
-        *uid += 1;
-        // ALLOW: id is created by inserting above, we also have no other way to access, thanks petgraph
-        let op = pipe_graph[id].to_op(*uid, supported_operators, None, None, None)?;
-        pipe_ops.insert(id, op);
-        inputs.insert(IN, id);
-
-        // TODO compute public streams - do not hard code
-        let id = pipe_graph.add_node(NodeConfig {
-            id: ERR,
-            kind: NodeKind::Output,
-            op_type: "passthrough".to_string(),
-            ..NodeConfig::default()
-        });
-        nodes.insert(ERR, id);
-        *uid += 1;
-        // ALLOW: id is created by inserting above, we also have no other way to access, thanks petgraph
-        let op = pipe_graph[id].to_op(*uid, supported_operators, None, None, None)?;
-        pipe_ops.insert(id, op);
-        outputs.push(id);
-
-        // TODO compute public streams - do not hardcode
-        let id = pipe_graph.add_node(NodeConfig {
-            id: OUT,
-            kind: NodeKind::Output,
-            op_type: "passthrough".to_string(),
-            ..NodeConfig::default()
-        });
-        nodes.insert(OUT, id);
-        *uid += 1;
-        // ALLOW: id is created by inserting above, we also have no other way to access, thanks petgraph
-        let op = pipe_graph[id].to_op(*uid, supported_operators, None, None, None)?;
-        pipe_ops.insert(id, op);
-        outputs.push(id);
-        let mut port_indexes: PortIndexMap = HashMap::new();
+        for (name, node_kind) in &BUILTIN_NODES {
```

Comment:
I am not familiar enough with the DAG form yet, but is it possible to be a problem that we insert the `METRICS` node here already instead of after processing the query, as we did before?

Reply:
From the perspective of query processing here, we are adding each node independently of another so the order of metrics node addition should not matter. I checked if there's anything in the way the metrics node is accessed later (as part of a running pipeline), that would benefit it from being inserted at the end, and I couldn't find anything suggesting that, so I think this move should be safe to do.